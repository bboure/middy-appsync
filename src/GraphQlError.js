export class GraphQlError extends Error {
  constructor (message, type = 'UnknownError', data = null, info = null) {
    super(message)
    this.type = type
    this.data = data
    this.info = info
  }
}
