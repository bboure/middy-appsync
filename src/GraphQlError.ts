export class GraphQlError extends Error {
  type: string;
  data?: unknown;
  info?: unknown;

  constructor(
    message: string,
    type = 'UnknownError',
    data = null,
    info = null,
  ) {
    super(message);
    this.type = type;
    this.data = data;
    this.info = info;
  }
}
