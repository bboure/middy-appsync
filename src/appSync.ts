import { AppSyncError } from './Errors';
import middy from '@middy/core';

type AppSyncResponse = {
  data?: unknown;
  errorInfo?: unknown;
  errorType?: string;
  errorMessage?: string;
};

const buildResponse = (response: unknown): AppSyncResponse => {
  if (response instanceof AppSyncError) {
    return {
      data: response.data,
      errorInfo: response.info,
      errorType: response.type,
      errorMessage: response.message,
    };
  } else if (response instanceof Error) {
    // re-throw returned errors
    throw response;
  } else {
    return { data: response };
  }
};

export const appSync: middy.Middleware<unknown, unknown> = () => {
  return {
    onError: async (handler) => {
      const response = buildResponse(handler.error);
      if (Array.isArray(handler.event)) {
        const resp = new Array(handler.event.length);
        handler.response = resp.fill(response);
      } else {
        handler.response = response;
      }

      // mark the error as handled
      return false;
    },

    after: async (handler) => {
      const { response } = handler;
      if (Array.isArray(handler.event)) {
        if (
          !Array.isArray(response) ||
          handler.event.length !== response.length
        ) {
          throw new Error(
            'BatchInvoke: The response does not match the request payload',
          );
        }

        handler.response = response.map(buildResponse);
      } else {
        handler.response = buildResponse(response);
      }
    },
  };
};
