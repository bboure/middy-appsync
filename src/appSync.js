import { GraphQlError } from './GraphQlError'
import { constant, times } from 'lodash'

const buildResponse = (response) => {
  if (response instanceof GraphQlError) {
    return {
      data: response.data,
      errorInfo: response.info,
      errorType: response.type,
      errorMessage: response.message
    }
  } else if (response instanceof Error) {
    return { errorMessage: 'Internal Server Error' }
  } else {
    return { data: response }
  }
}

export function appSync () {
  return {
    onError: async (handler, next) => {
      if (handler.error instanceof GraphQlError) {
        const response = buildResponse(handler.error)
        if (Array.isArray(handler.event)) {
          handler.response = times(handler.event.length, constant(response))
        } else {
          handler.response = response
        }

        return false
      }

      return handler.error
    },

    after: async (handler) => {
      const { response } = handler

      if (Array.isArray(handler.event)) {
        if (!Array.isArray(response) || handler.event.length !== response.length) {
          throw new Error('BatchInvoke: The response does not match the request payload')
        }

        handler.response = response.map(buildResponse)
      } else {
        handler.response = buildResponse(response)
      }
    }
  }
}
