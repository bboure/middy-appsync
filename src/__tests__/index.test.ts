import middy from '@middy/core';
import { Context } from 'aws-lambda';
import { appSync } from '../appSync';
import { GraphQlError } from '../GraphQlError';

const fakeConext: Context = ({} as unknown) as Context;

describe('appsync middleware test suite', () => {
  it('Should wrap the response in a GraphQl template', () => {
    const handler = middy((event, context, cb) => {
      cb(null, {
        field1: 'foo',
        field2: 'bar',
      });
    });

    handler.use(appSync());

    handler({}, fakeConext, (_, response) => {
      expect(response).toMatchSnapshot();
    });
  });

  it('Should handle a GraphQlError error', () => {
    const handler = middy((event, context, cb) => {
      cb(
        new GraphQlError(
          'Error message',
          'Error',
          { some: 'data' },
          { info: 'value' },
        ),
      );
    });

    handler.use(appSync());

    handler({}, fakeConext, (_, response) => {
      expect(response).toMatchSnapshot();
    });
  });

  it('Should handle a thrown GraphQlError', () => {
    const handler = middy(() => {
      throw new GraphQlError(
        'Thrown Error message',
        'Thrown Error',
        { some: 'data' },
        { info: 'value' },
      );
    });

    handler.use(appSync());

    handler({}, fakeConext, (_, response) => {
      expect(response).toMatchSnapshot();
    });
  });

  it('Should handle a GraphQlError response', () => {
    const handler = middy((_event, context, cb) => {
      cb(
        null,
        new GraphQlError(
          'Returned Error message',
          'ReturnedError',
          { some: 'data' },
          { info: 'value' },
        ),
      );
    });

    handler.use(appSync());

    handler({}, fakeConext, (_, response) => {
      expect(response).toMatchSnapshot();
    });
  });

  it('Should handle a thrown GraphQlError with async', () => {
    const handler = middy(
      async (): Promise<unknown> => {
        throw new GraphQlError(
          'Thrown Error message',
          'Thrown Error',
          { some: 'data' },
          { info: 'value' },
        );
      },
    );

    handler.use(appSync());

    handler({}, fakeConext, (_, response) => {
      expect(response).toMatchSnapshot();
    });
  });

  it('Should handle a returned GraphQlError with async', () => {
    const handler = middy(async () => {
      return new GraphQlError(
        'Returned Error message',
        'ReturnedError',
        { some: 'data' },
        { info: 'value' },
      );
    });

    handler.use(appSync());

    handler({}, fakeConext, (_, response) => {
      expect(response).toMatchSnapshot();
    });
  });

  it('Should handle a standard Error anonymously', () => {
    const handler = middy((event, context, cb) => {
      cb(new Error('Some Error'));
    });

    handler.use(appSync());

    handler({}, fakeConext, (error, response) => {
      expect(error).toBeNull();
      expect(response.errorMessage).toEqual('Internal Server Error');
    });
  });

  it('Should handle a standard Error response anonymously', () => {
    const handler = middy((event, context, cb) => {
      cb(
        null,
        new Error(
          'Uncaught ReferenceError: myVar is not defined at index.js:123:456',
        ),
      );
    });

    handler.use(appSync());

    handler({}, fakeConext, (error, response) => {
      expect(error).toBeNull();
      expect(response.errorMessage).toEqual('Internal Server Error');
    });
  });

  it('Should handle a standard Error anonymously', () => {
    const handler = middy((event, context, cb) => {
      cb(
        new Error(
          'Uncaught ReferenceError: myVar is not defined at index.js:123:456',
        ),
      );
    });

    handler.use(appSync());

    handler({}, fakeConext, (error, response) => {
      expect(error).toBeNull();
      expect(response.errorMessage).toEqual('Internal Server Error');
    });
  });

  it('Should succeed when response matches event in batches', () => {
    const handler = middy((event, context, cb) =>
      cb(null, [{ foo: 'bar' }, { biz: 'baz' }]),
    );
    handler.use(appSync());
    handler([{}, {}], fakeConext, (_, message) => {
      expect(message).toMatchSnapshot();
    });
  });

  it('Should accept mixed errors/responses in batches', () => {
    const handler = middy((event, context, cb) =>
      cb(null, [{ foo: 'bar' }, new GraphQlError('Not Found', 'NotFound')]),
    );
    handler.use(appSync());
    handler([{}, {}], fakeConext, (_, message) => {
      expect(message).toMatchSnapshot();
    });
  });

  it('Should reject the whole batch when returning an error in batch', () => {
    const handler = middy((event, context, cb) => {
      cb(new GraphQlError('Internal Error', 'NotFoundInternal Error'));
    });
    handler.use(appSync());
    handler([{}, {}], fakeConext, (_, message) => {
      expect(message).toMatchSnapshot();
    });
  });

  it('Should reject the whole batch when throwing an error in batch', () => {
    const handler = middy(() => {
      throw new GraphQlError('Internal Error', 'Internal Error');
    });
    handler.use(appSync());
    handler([{}, {}], fakeConext, (_, message) => {
      expect(message).toMatchSnapshot();
    });
  });

  it('Should reject the whole batch when throwing an error in batch with async', () => {
    const handler = middy(
      async (): Promise<unknown> => {
        throw new GraphQlError('Internal Error', 'Internal Error');
      },
    );
    handler.use(appSync());
    handler([{}, {}], fakeConext, (_, message) => {
      expect(message).toMatchSnapshot();
    });
  });

  it('Should fail when the response is not an array but the event is', () => {
    const handler = middy((event, context, cb) => {
      cb(null, { foo: 'bar' });
    });

    const middleware = appSync();
    const spy = jest.spyOn(middleware, 'after');

    handler.use(middleware);

    handler([{}, fakeConext, {}], fakeConext, async () => {
      expect(spy).toHaveBeenCalledTimes(1);
      await expect(spy.mock.results[0].value).rejects.toMatchSnapshot();
    });
  });

  it('Should fail when the response length does not match the event length', () => {
    const handler = middy((event, context, cb) => {
      cb(null, ['foo', 'bar']);
    });

    const middleware = appSync();
    const spy = jest.spyOn(middleware, 'after');

    handler.use(middleware);

    handler([{}, fakeConext, {}], fakeConext, async () => {
      expect(spy).toHaveBeenCalledTimes(1);
      await expect(spy.mock.results[0].value).rejects.toMatchSnapshot();
    });
  });
});
