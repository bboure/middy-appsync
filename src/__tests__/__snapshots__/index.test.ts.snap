// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middleware should accept mixed errors/responses in batches 1`] = `
Array [
  Object {
    "data": Object {
      "foo": "bar",
    },
  },
  Object {
    "data": null,
    "errorInfo": null,
    "errorMessage": "Not Found",
    "errorType": "NotFound",
  },
]
`;

exports[`middleware should fail when the response is not an array but the event is 1`] = `[Error: BatchInvoke: The response does not match the request payload]`;

exports[`middleware should fail when the response length does not match the event length 1`] = `[Error: BatchInvoke: The response does not match the request payload]`;

exports[`middleware should handle a NotFoundException 1`] = `
Object {
  "data": null,
  "errorInfo": null,
  "errorMessage": "Resource not found",
  "errorType": "NotFound",
}
`;

exports[`middleware should handle a returned AppSyncError with async 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "errorInfo": Object {
    "info": "value",
  },
  "errorMessage": "Returned Error message",
  "errorType": "ReturnedError",
}
`;

exports[`middleware should handle a thrown AppSyncError 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "errorInfo": Object {
    "info": "value",
  },
  "errorMessage": "Thrown Error message",
  "errorType": "Thrown Error",
}
`;

exports[`middleware should handle a thrown AppSyncError with async 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "errorInfo": Object {
    "info": "value",
  },
  "errorMessage": "Thrown Error message",
  "errorType": "Thrown Error",
}
`;

exports[`middleware should handle an AppSyncError 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "errorInfo": Object {
    "info": "value",
  },
  "errorMessage": "Error message",
  "errorType": "Error",
}
`;

exports[`middleware should handle an AppSyncError response 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "errorInfo": Object {
    "info": "value",
  },
  "errorMessage": "Returned Error message",
  "errorType": "ReturnedError",
}
`;

exports[`middleware should leave standard Errors 1`] = `[Error: Uncaught ReferenceError: myVar is not defined at index.js:123:456]`;

exports[`middleware should leave standard Errors 2`] = `undefined`;

exports[`middleware should re-throw standard Error 1`] = `[Error: Uncaught ReferenceError: myVar is not defined at index.js:123:456]`;

exports[`middleware should re-throw standard Error 2`] = `undefined`;

exports[`middleware should re-throw standard Error response 1`] = `[Error: Uncaught ReferenceError: myVar is not defined at index.js:123:456]`;

exports[`middleware should re-throw standard Error response 2`] = `undefined`;

exports[`middleware should reject the whole batch when returning an error 1`] = `
Array [
  Object {
    "data": null,
    "errorInfo": null,
    "errorMessage": "Internal Error",
    "errorType": "Internal Error",
  },
  Object {
    "data": null,
    "errorInfo": null,
    "errorMessage": "Internal Error",
    "errorType": "Internal Error",
  },
]
`;

exports[`middleware should reject the whole batch when throwing an error 1`] = `
Array [
  Object {
    "data": null,
    "errorInfo": null,
    "errorMessage": "Internal Error",
    "errorType": "Internal Error",
  },
  Object {
    "data": null,
    "errorInfo": null,
    "errorMessage": "Internal Error",
    "errorType": "Internal Error",
  },
]
`;

exports[`middleware should reject the whole batch when throwing an error with async 1`] = `
Array [
  Object {
    "data": null,
    "errorInfo": null,
    "errorMessage": "Internal Error",
    "errorType": "Internal Error",
  },
  Object {
    "data": null,
    "errorInfo": null,
    "errorMessage": "Internal Error",
    "errorType": "Internal Error",
  },
]
`;

exports[`middleware should succeed when response matches event in batches 1`] = `
Array [
  Object {
    "data": Object {
      "foo": "bar",
    },
  },
  Object {
    "data": Object {
      "biz": "baz",
    },
  },
]
`;

exports[`middleware should wrap the response in an AppSync response object 1`] = `
Object {
  "data": Object {
    "field1": "foo",
    "field2": "bar",
  },
}
`;
